docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+
+#### Individual tests
+
+Within a `describe` block, individual tests are encapsulated by `it` methods,
+which is an alias of
+[Jest’s `test` method](https://jestjs.io/docs/en/api#testname-fn-timeout).
+
+Each test must include one or more assertions (usually only one)
+that confirm that the result of executing some code is what you expected.
+These are called ["matchers"](https://jestjs.io/docs/en/using-matchers)
+in Jest parlance, and they all follow this format:
+
+```js
+expect( someValue ).someKindOfComparisonWith( someOtherValue );
+```
+
+For example, let's take another look at the sample tests we wrote above:
+
+```js
+describe( 'sample', () => {
+
+    describe( 'gimmeString()', () => {
+
+      it( 'should return a string with expected value', () => {
+        sampleString = 'Shredder';
+        expect( sample.gimmeString() ).toBe( sampleString );
+      } );
+
+    } );
+
+    describe( 'gimmeObject()', () => {
+
+      it( 'should return an object with expected value', () => {
+        sampleObject = {
+            image: 'https://vignette.wikia.nocookie.net/tmnt/images/0/00/Krangnobody.png',
+            caption: 'Krang portrait'
+        };
+        expect( sample.gimmeObject() ).toBe( sampleObject );
+      } );
+
+    } );
+
+} );
+```
+
+In these tests, we check _(`expect`)_ that
+the string returned by `sample.gimmeString()`
+and the object returned by `sample.gimmeObject()`
+match _(`toBe`)_ the `sampleString` and `sampleObject`
+that we defined in the tests.
+
+There are many kinds of assertions you can use
+besides the simple equality comparison of `toBe`.
+See [the Jest Using Matchers guide](https://jestjs.io/docs/en/using-matchers)
+for a primer on them and
+[the Jest `expect` API docs](https://jestjs.io/docs/en/expect)
+for the full list of its matchers.
+
+
+### Providing test data
+
+The first principle of test data for unit tests is to
+**keep test data as simple as possible** –
+use the minimum needed to test the code.
+
+#### Direct definition of test data
+
+The simplest way to set up test data is to
+declare it as variables within each test, e.g., the tests in
+[strings-spec.js](https://github.com/cfpb/cfgov-refresh/blob/master/test/unit_tests/js/modules/util/strings-spec.js).
+This can include HTML markup for DOM manipulation tests,
+if each test requires different markup.
+
+#### Setup and teardown methods
+
+If you will need to leverage the same test data across different tests,
+Jest has setup and teardown methods, such as
+[`beforeEach` and `afterEach`](https://jestjs.io/docs/en/setup-teardown#repeating-setup-for-many-tests), or [`beforeAll` and `afterAll`](https://jestjs.io/docs/en/setup-teardown#one-time-setup),
+which can be used to performing actions
+that are needed before and after running all tests or each test in a suite.
+For example, the tests in
+[`Analytics-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/modules/Analytics-spec.js#L7-L32)
```

Comment:
This links to this PR branch, which might be deleted in future. Link to a specific commit SHA instead?